### ISSUE TITLE
[SOLVED] Loss won't decrease, predictions are all the same

### ISSUE THREAD

**Issue by MarcoBauzz** [shekkizh/FCN.tensorflow#1]
Hi! I would like to reproduce your results.
Just running the code like `python FCN.py` doesn't seem to do the job for me.
The default parameters are:
- IMAGE_SIZE = 224 ~~(changing it to 256 does not affect the results)~~
- learning_rate = 1e-4
- batch_size = 2

What I get is that training and validation loss start at about 400, and very quickly (200 iterations) decrease until they settle to about 3.

```
Step: 0, Train_loss:415.754
2016-10-13 12:19:13.407670 ---> Validation_loss: 395.876
Step: 10, Train_loss:28.7208
Step: 20, Train_loss:10.2944
Step: 30, Train_loss:5.06159
Step: 40, Train_loss:4.51668
Step: 50, Train_loss:4.17936
Step: 60, Train_loss:4.55051
Step: 70, Train_loss:4.98752
Step: 80, Train_loss:3.63942
Step: 90, Train_loss:3.56676
Step: 100, Train_loss:3.96641
Step: 110, Train_loss:3.72767
Step: 120, Train_loss:3.26587
Step: 130, Train_loss:3.89015
Step: 140, Train_loss:5.48371
Step: 150, Train_loss:4.27173
Step: 160, Train_loss:3.81378
Step: 170, Train_loss:3.58391
Step: 180, Train_loss:2.79207
Step: 190, Train_loss:4.10269
Step: 200, Train_loss:4.57686
Step: 210, Train_loss:4.00551
Step: 220, Train_loss:3.1667
Step: 230, Train_loss:3.7841
Step: 240, Train_loss:3.74983
Step: 250, Train_loss:3.03212
Step: 260, Train_loss:2.85248
Step: 270, Train_loss:3.64257
Step: 280, Train_loss:3.765
Step: 290, Train_loss:4.16679
Step: 300, Train_loss:4.0291
Step: 310, Train_loss:3.95092
Step: 320, Train_loss:3.38709
Step: 330, Train_loss:2.48646
Step: 340, Train_loss:2.98015
Step: 350, Train_loss:3.59501
Step: 360, Train_loss:3.80755
Step: 370, Train_loss:3.73314
Step: 380, Train_loss:3.40185
Step: 390, Train_loss:3.89394
Step: 400, Train_loss:3.80676
Step: 410, Train_loss:2.78324
Step: 420, Train_loss:3.14695
Step: 430, Train_loss:3.29019
Step: 440, Train_loss:3.16163
Step: 450, Train_loss:3.64598
Step: 460, Train_loss:2.74009
Step: 470, Train_loss:3.93917
Step: 480, Train_loss:3.815
Step: 490, Train_loss:3.83076
Step: 500, Train_loss:4.45192
2016-10-13 12:24:10.606606 ---> Validation_loss: 3.02666
```

I kept it running up to 35000 iterations, which should be about 3.5 epochs, but the loss won't decrease any further.
If I then validate the model at 35000 iterations (Train_loss:2.73392, Validation_loss: 3.51286) with `python FCN.py --mode visualize` I get always the same prediction, **whichever the input image is**:

![image](https://cloud.githubusercontent.com/assets/20168067/19381691/2066e62a-91fd-11e6-8358-39e6ec15d05f.png)

This is also, by the way, the same prediction I get with an earlier model (200 iterations).

Is there something I'm getting wrong?
Thank you

**Comment by shekkizh** [comment]
Hi,
 500 iterations is too small for the network to learn.  One way to figure out what might be wrong is to look at the gradients summary in tensorboard. Train with debug set to True, `python FCN.py --debug=True` and looking at the tensorboard graphs.
Given that you ran for 35k iterations I would think the network is stuck at a plateau or the Adam optimizer is slowing the training. Try **restarting the training with the saved model** or lower the beta in Adam optimizer. I was able to see some results around 7000 iterations.

**Comment by MarcoBauzz** [comment]
Thanks for the quick answer.
I ran the script again, this time with
- IMAGE_SIZE = 256
- learning_rage = 5e-5
- saving checkpoint every 10000 iterations (doing this every 500 took away a lot of time)

Checking the result at 35k iterations now produces meaningful predictions (I will keep it running even further now).

It is possible that I messed up with the checkpoints the first time, since I had to stop and resume the computation.
Or the two parameters I changed really had an effect.
Or I was subject to an unlucky initialization the first time.

Anyway, thanks again for the support!

**Comment by ayush1997** [comment]
Hi @MarcoBauzz 
How did you get around this issue I am facing the exact issue of same result image for every test image after 24k iterations.
IMAGE_SIZE = 224
learning_rage = 1e-4

Thanks!

**Comment by MarcoBauzz** [comment]
Hi, as I mentioned I don't know exactly the reason why it started working.
The only thing I could do was reporting any change I made to the code/settings.

Also, the point made by @shekkizh about the plateau and/or the optimizer makes sense, even though (in my case) it wasn't needed for the problem to get solved.

**Comment by ayush1997** [comment]
Thanks @MarcoBauzz  !

**Comment by ayush1997** [comment]
Hi @shekkizh 
Does the code restores training from the saved model if it is run(`python FCN.py`) again.

Thanks!

**Comment by shekkizh** [comment]
Yes. The code always looks for saved model in "logs_dir" and restores the saved model.

**Comment by ayush1997** [comment]
Thanks for a quick reply.I am not sure if it keeps track of images that are trained already, as on running again it starts saving model again from 0 iteration(model.ckpt-0.meta) onwards.
Though I guess that wont matter much if it restores pretrained model and even trains again on same images, what do you think.

**Comment by varungupta31** [comment]
Hey @MarcoBauzz, I understand it's been a while since you used this repo, help me out if you can.
I'm trying to train this model for 2 classes, and trying to visaulize the training metrics like - accuracy, mean accuracy, mean iou.f.w iou, etc. When you trained the model with `debug = True` were you getting all these curves in tensorboard?

Thanks.
